File: packages/ts/src/client/rdk-client.ts
/**
 * `RdkClient` — the high-level entry point. Resolves a network, composes the
 * REST and `qor_` JSON-RPC read clients, exposes the QCAI-assisted profile
 * suggestion, and connects a signing tx client.
 */
import type { OfflineSigner } from "@cosmjs/proto-signing";
import type { NetworkName } from "../constants";
import { getNetwork, type Endpoints, type NetworkConfig } from "../config/networks";
import { suggestProfile, type ProfileSuggestion } from "../profiles/suggest";
import { RdkTxClient, type RdkTxClientConnectOptions } from "../tx/client";
import { QorClient } from "./jsonrpc";
import { RestClient } from "./rest";
import type { FetchLike } from "./http";
import type { ParamsView } from "./views";
import type { ProfileName } from "../config/enums";

export interface CreateRdkClientOptions {
  /** Network preset (default `testnet`). */
  network?: NetworkName;
  /** Endpoint overrides; merged onto the network preset's defaults. */
  endpoints?: Partial<Endpoints>;
  /** Custom fetch (for testing or non-standard environments). */
  fetch?: FetchLike;
}

export class RdkClient {
  /** The resolved network (chain id and endpoints). */
  readonly network: NetworkConfig;
  /** REST (LCD) read client. */
  readonly rest: RestClient;
  /** `qor_` JSON-RPC client. */
  readonly qor: QorClient;

  constructor(options: CreateRdkClientOptions = {}) {
    const net = getNetwork(options.network ?? "testnet");
    net.endpoints = { ...net.endpoints, ...options.endpoints };
    this.network = net;
    this.rest = new RestClient(net.endpoints.rest, { fetch: options.fetch });
    this.qor = new QorClient(net.endpoints.evmRpc, { fetch: options.fetch });
  }

  /** Read the live `rdk` module parameters from the chain. */
  params(): Promise<ParamsView> {
    return this.rest.getParams();
  }

  /** QCAI-assisted profile suggestion, with a documented fallback to `defi`. */
  suggestProfile(useCase: string, opts?: { fallback?: ProfileName }): Promise<ProfileSuggestion> {
    return suggestProfile(useCase, this.qor, opts);
  }

  /** Connect a signing tx client at the consensus RPC endpoint. */
  connectTx(signer: OfflineSigner, opts?: RdkTxClientConnectOptions): Promise<RdkTxClient> {
    return RdkTxClient.connect(this.network.endpoints.rpc, signer, opts);
  }
}

/** Create an {@link RdkClient}. Defaults to the public testnet. */
export function createRdkClient(options: CreateRdkClientOptions = {}): RdkClient {
  return new RdkClient(options);
}

File: packages/py/src/qorrdk/client/rdk_client.py
"""``RdkClient`` -- the high-level entry point.

Resolves a network, composes the REST and ``qor_`` JSON-RPC read clients, exposes
the QCAI-assisted profile suggestion, and connects a signing tx client.
"""

from __future__ import annotations

from typing import Optional

from ..config.networks import NetworkConfig, get_network
from ..enums import ProfileName
from .http import Transport
from .jsonrpc import QorClient
from .rest import RestClient
from .views import ParamsView


class RdkClient:
    """The high-level RDK facade over a resolved network."""

    def __init__(
        self,
        network: str = "testnet",
        endpoints: Optional[dict] = None,
        transport: Optional[Transport] = None,
    ) -> None:
        net = get_network(network)
        if endpoints:
            for key, value in endpoints.items():
                if not hasattr(net.endpoints, key):
                    raise ValueError(f'unknown endpoint "{key}"')
                setattr(net.endpoints, key, value)
        #: The resolved network (chain id and endpoints).
        self.network: NetworkConfig = net
        #: REST (LCD) read client.
        self.rest = RestClient(net.endpoints.rest, transport=transport)
        #: ``qor_`` JSON-RPC client.
        self.qor = QorClient(net.endpoints.evm_rpc, transport=transport)
        self._transport = transport

    def params(self) -> ParamsView:
        """Read the live ``rdk`` module parameters from the chain."""
        return self.rest.get_params()

    def suggest_profile(
        self, use_case: str, fallback: ProfileName = ProfileName.DEFI
    ):
        """QCAI-assisted profile suggestion, with a documented fallback to ``defi``."""
        from ..profiles import suggest_profile

        return suggest_profile(use_case, self.qor, fallback)

    def connect_tx(self, signer) -> "RdkTxClient":
        """Connect a signing tx client bound to this network's REST + chain id."""
        from ..tx.client import RdkTxClient

        return RdkTxClient(
            rest_url=self.network.endpoints.rest,
            chain_id=self.network.chain_id,
            signer=signer,
            transport=self._transport,
        )


def create_rdk_client(
    network: str = "testnet",
    endpoints: Optional[dict] = None,
    transport: Optional[Transport] = None,
) -> RdkClient:
    """Create an :class:`RdkClient`. Defaults to the public testnet."""
    return RdkClient(network=network, endpoints=endpoints, transport=transport)


__all__ = ["RdkClient", "create_rdk_client"]
